File: PaperLib/Injected/AppState.swift
//
//  AppState.swift
//  PaperLib
//
//  Created by GeoffreyChen on 26/11/2021.
//

import Combine
import RealmSwift
import SwiftUI

class StateWrapper<T: Equatable> {
    private var _state: T
    private var _lastState: T

    var value: T {
        get {
            return self._state
        }
        set {
            self._state = newValue
            if newValue != self._lastState {
                self._lastState = newValue
                self.publisher.send(newValue)
            }
        }
    }
    var binding: Binding<T> {
        Binding<T>(
            get: {
                return self._state
            },
            set: {
                self._state = $0
                if $0 != self._lastState {
                    self._lastState = $0
                    self.publisher.send($0)
                }
            }
        )
    }
    let publisher: CurrentValueSubject<T, Never>

    init(_ initState: T) {
        _state = initState
        _lastState = initState
        self.publisher = CurrentValueSubject<T, Never>(_state)
    }

}

struct SharedState {
    var selection = SelectionState()
    var viewState = ViewState()
    var sharedData = SharedData()
}

// MARK: -

struct SelectionState {
    var selectedCategorizer = StateWrapper(nil as String?)
    var selectedIds = StateWrapper(Set<ObjectId>.init())
}

struct ViewState {
    var entitiesCount = StateWrapper(0)

    var entitiesViewSwitcher = StateWrapper(0)
    var entitiesViewSortSwitcher = StateWrapper(nil as String?)

    var isEditViewShown = StateWrapper(false)
    var isTagViewShown = StateWrapper(false)
    var isFolderViewShown = StateWrapper(false)
    var isNoteViewShown = StateWrapper(false)

    var alertInformation = StateWrapper(nil as String?)

    var processingQueueCount = StateWrapper(0)
    var realmReinited = StateWrapper(nil as Date?)

    var searchMode = StateWrapper(SearchMode.general)
}

struct SharedData {
    var searchQuery = StateWrapper(nil as String?)
    var editEntityDraft = StateWrapper(PaperEntityDraft())
}

File: src/interactors/app-state.ts
import { v4 as uuid } from 'uuid';
import { app, BrowserWindow } from '@electron/remote';

export class State<T> {
  value: T;
  id: string;
  checkDuplicated: boolean;

  constructor(value: T, checkDuplicated = true) {
    this.value = value;
    this.id = uuid();
    this.checkDuplicated = checkDuplicated;
  }

  get() {
    return this.value;
  }

  set(value: T) {
    if (value === this.value && this.checkDuplicated) {
      return;
    }
    this.value = value;
    try {
      const window = BrowserWindow.getAllWindows()[0];
      if (window) {
        window.webContents.send(this.id, this.value);
      }
    } catch (e) {
      console.log(e);
    }
  }
}

export class SharedState {
  viewState: Record<string, State<number | string | boolean>>;
  selectionState: Record<string, State<number | string | boolean>>;
  sharedData: Record<string, State<number | string | boolean>>;
  dbState: Record<string, State<number | string | boolean>>;

  constructor() {
    this.viewState = {
      processingQueueCount: new State(0, false),
      entitiesCount: new State(0),

      sortBy: new State('addTime'),
      sortOrder: new State('desc'),
      searchText: new State(''),
      searchMode: new State('general'),

      isEditViewShown: new State(false),
      isTagViewShown: new State(false),
      isFolderViewShown: new State(false),
      isNoteViewShown: new State(false),
      isPreferenceViewShown: new State(false),

      preferenceUpdated: new State(new Date().getTime()),
      themeUpdated: new State(new Date().getTime()),
      realmReinited: new State(new Date().getTime()),

      alertInformation: new State('', false),

      viewType: new State('list'),
      theme: new State('light'),
      syncFileStorageAvaliable: new State(false),
    };

    // Selection State
    this.selectionState = {
      selectedIndex: new State(''),
      selectedCategorizer: new State(''),
    };

    // Shared Data
    this.sharedData = {
      editEntityDraft: new State(''),
    };
    // DB State
    this.dbState = {
      entitiesUpdated: new State(new Date().getTime()),
      tagsUpdated: new State(new Date().getTime()),
      foldersUpdated: new State(new Date().getTime()),
      defaultPath: new State(app.getPath('userData')),
      selectedPath: new State(''),
    };
  }

  get(dest: string) {
    const state = getObj(this, dest) as State<number | string | boolean>;
    return state;
  }

  set(dest: string, value: number | string | boolean) {
    const state = getObj(this, dest) as State<number | string | boolean>;
    state.set(value);
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getObj(obj: any, dest: string) {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-member-access
  return dest.split('.').reduce((a, b) => a[b], obj);
}
